com.wamas.ide.launching.ui/src/com/wamas/ide/launching/ui/build/LcDslTargetPlatformSupport.java

```diff
@@ -1,664 +1,666 @@
+        }
+    }
+
+    private Map<URI, URI> all(ModelEntry entry, SortedMap<URI, URI> map) {
+        IPluginModelBase model = entry.getModel();
+        String id = entry.getId();
+        URI low;
+        if (model != null) {
+            String version = model.getBundleDescription().getVersion().toString();
+            low = URI.createURI("target:/" + id + "/" + version + "/", true);
+        } else {
+            low = URI.createURI("target:/" + id + "/", true);
+        }
+
+        URI high = low.trimSegments(1).appendSegment(String.valueOf(Character.MAX_VALUE));
+        return map.subMap(low, high);
+    }
+
+    @Override
+    public void removeProject(ToBeBuilt toBeBuilt, IProject project, IProgressMonitor monitor) {
+        // Looks like a clean build was triggered - assume that we want to rebuild the
+        // configurations from the TP. Make sure that we still keep the URIs from the TP known
+        // to the Xtext index.
+        SortedMap<URI, URI> map = Futures.getUnchecked(uriMap);
+        lock.writeLock().lock();
+        try {
+            pendingUpdated.addAll(map.keySet());
+            toBeBuilt.getToBeDeleted().removeAll(map.keySet());
+        } finally {
+            lock.writeLock().unlock();
+        }
+    }
+
+    @Override
+    public void updateProject(ToBeBuilt toBeBuilt, IProject project, IProgressMonitor monitor) throws CoreException {
+        // A full build was triggered for a project
+        SortedMap<URI, URI> map = Futures.getUnchecked(uriMap);
+        lock.writeLock().lock();
+        try {
+            // Make sure that we don't remove target:/ URIs automagically from the Xtext index
+            // but only if they are no longer managed from here
+            toBeBuilt.getToBeDeleted().removeAll(Sets.filter(map.keySet(), uri -> !project.getName().equals(uri.segment(1))));
+        } finally {
+            lock.writeLock().unlock();
+        }
+    }
+
+    @Override
+    public boolean removeStorage(ToBeBuilt toBeBuilt, IStorage storage, IProgressMonitor monitor) {
+        // nothing to do
+        return false;
+    }
+
+    @Override
+    public boolean updateStorage(ToBeBuilt toBeBuilt, IStorage storage, IProgressMonitor monitor) {
+        // nothing to do
+        return false;
+    }
+
+    @Override
+    public boolean isPossiblyHandled(IStorage storage) {
+        // nothing to do
+        return false;
+    }
+
+    @Override
+    public void stateResolved(StateDelta delta) {
+        // nothing to do
+    }
+
+    @Override
+    public void stateChanged(State newState) {
+        TreeMap<URI, URI> newContent = Arrays
+                .stream(PDECore.getDefault().getModelManager().getExternalModelManager().getAllModels()).parallel()
+                .flatMap(LcDslTargetPlatformSupport::added).collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue,
+                        this::error, LcDslTargetPlatformSupport::createSortedUriMap));
+        if (!uriMap.isDone()) {

```

Comment:
I think this can be moved above line 549 to avoid unnecessary work.

Reply:
Race condition :) Approved.

Reply:
Yes, it can.
It might make sense to register the listeners (line 198) only after the map was computed (line 207).

Reply:
ok, panic mode: i tried to handle all possible usages with precaution.

Reply:
Yes, makes sense. It might need refinements to close the gap between the first query of the external model and interleaved events that won't be seen right now, but these *should* be rather unlikely.